[plugins.telesync] auto-restrict new users

deffnull opened the pull request:
- can be controlled with per conversation config key “sync_restrict_user”, default off

deffnull committed:
new config key for auto-restrict

deffnull committed:
[plugins.telesync] auto-restrict new users

- can be controlled with new per conversation config key “sync_restrict_user”

das7pad commented on hangupsbot/sync/__init__.py:
```diff
@@ -131,6 +131,10 @@
+
+    # automatically restrict TG users on join, possible values are: messages,
+    # media, sticker, websites, sticker+websites
+    'sync_restrict_user': '',
```
Let's keep the telesync defaults in `plugins.telesync.setup_config.default_config` and access them with `bot.get_config_suboption`.

das7pad commented on hangupsbot/plugins/telesync/core.py:
```diff
@@ -557,6 +566,34 @@ def _on_membership_change(self, msg):
+                          NO_WEBPREVIEW_AND_STICKER_RIGHTS if restrict == 'sticker+websites'
+                          else None)
+                if rights and msg['chat']['type'] == 'supergroup':
+                    result = await self.restrictChatMember(msg.chat_id,
```
We are actually not in a coroutine here, this requires `async def _on_membership_change(...):` and `await self._on_membership_change(msg)` in line 485.

das7pad replied:
`TelegramBot.restrictChatMember` likely raises an Exception - the membership change would never reach the sync handler in this case.

das7pad commented on hangupsbot/plugins/telesync/core.py:
```diff
@@ -557,6 +566,34 @@ def _on_membership_change(self, msg):
+                                                           **rights)
+
+                    if result is not True:
+                        mod_chat = self.config('mod_chat') if 'mod_chat' in self.config() else None
```
What about an entry in `plugins.telesync.setup_config.default_config` for `mod_chat`?

das7pad reviewed (changes requested):
Great idea to add another mod function to `telesync`!

There is a lot of duplicate code here: what about splitting `plugins.telesync.commands_tg.command_restrict_user` into a core and command head:
- the head would enforce an admin user and `args`/valid option and sends the result back
- the core is feed with the `restrict` mode and `args` to parse the specified user_ids; it could simply return the `failed` dict.

We could use the restrict_user core then to run it with the user ids of the changed member.

das7pad commented:
> what about splitting plugins.telesync.commands_tg.command_restrict_user into a core and command head

done in https://github.com/das7pad/hangoutsbot/commit/71b6ca0d7e2eb7d8af25c4223773a0936eb16677

deffnull committed:
Revert "new config key for auto-restrict"

This reverts commit a73cf335918da30a91da2e9cf7e65f1402cfddce.

deffnull committed:
Revert "[plugins.telesync] auto-restrict new users"

This reverts commit 539c4bf9f783d2475f4f04ff4f1dba33b818bcab.

deffnull committed:
Merge remote-tracking branch 'das7pad/v3.1-das7pad' into feature/auto-restrict

deffnull committed:
[plugins.telesync] new default config mod_chat

can point to a TG chat with mods to be alerted about important events

deffnull committed:
[plugins.telesync] new config option restrict_users

default: ‘’, possible values are: messages, media, sticker, websites, sticker+websites

deffnull committed:
[plugins.telesync] auto-restrict new chat members

feature is based on the per TG-conv setting ‘restrict_users’

deffnull commented:
Changes applied based on 71b6ca0

das7pad commented:
Thank you!
